src/main/java/com/aws/greengrass/status/FleetStatusService.java

```diff
@@ -568,10 +617,27 @@ private void uploadFleetStatusServiceData(Set<GreengrassService> greengrassServi
                 .messageType(MessageType.fromTrigger(trigger))
                 .deploymentInformation(deploymentInformation)
                 .build();
-
-        publisher.publish(fleetStatusDetails, components);
-        logger.atInfo().event("fss-status-update-published").kv("trigger", trigger)
-                .log("Status update published to FSS");
+        synchronized (this) {
+            // make sure only one thread updates lastFSSPublishTime and executes publish function
+            // make sure messages are not published too frequently to cause message drop from cloud
+            if (lastFSSPublishTime.plusSeconds(FLEET_STATUS_MESSAGE_PUBLISH_MAX_WAIT_TIME_SEC)
+                    .isBefore(Instant.now())) {
+                publisher.publish(fleetStatusDetails, components);
+                lastFSSPublishTime = Instant.now();
+                logger.atInfo().event("fss-status-update-published").kv("trigger", trigger)
+                        .log("Status update published to FSS");
+            } else {
+                // Calculate the time that next message needs to wait.
+                Duration duration = Duration.between(Instant.now(), lastFSSPublishTime
+                        .plusSeconds(FLEET_STATUS_MESSAGE_PUBLISH_MAX_WAIT_TIME_SEC));
+                lastFSSPublishTime = Instant.now().plusSeconds(duration.getSeconds());
+                ses.schedule(() -> {
```

Comment:
Do you have a test scenario that spams FSS messages and make sure this publish happens every x seconds?